### ISSUE TITLE
Currently open filenames not syncing

### ISSUE THREAD

**Issue by russell-kitchen** [iina/iina#3159]
**System and IINA version:**

- macOS 10.15.7 (19H2)
- IINA 1.1.0 Build 123

**Expected behavior:**
Currently open items list to reflect the currently open items.
<img width="271" alt="Screenshot+2020-10-16+06 21 59" src="https://user-images.githubusercontent.com/11626556/96216372-0b006700-0f78-11eb-8f04-3db56b1ec779.png">

**Actual behavior:**
Currently open items list does not sync as intended. When I close a window and open a new one, the title of the new file in the list is that of the old one. So for example, if I close A.mp4 then open B.mp4, in the currently open list the freshly opened B.mp4 will incorrectly have the title A.mp4. Similarly if I have a playlist running, the filenames do not update in the current list when moving to the next or previous.

**Steps to reproduce:**
Close video window with the top left red button or with shortcut, then open another window.

**How often does this happen?**
Every time I close and open a window.

**Comment by low-batt** [comment]
I too have been hitting this issue.

The AppKit framework is responsible for maintaining the list of windows displayed by the "Window" menu. In `PlayerCore` IINA is caching and reusing players which includes their `NSWindow` objects. For some reason the reuse is causing the framework to fail to update the window title displayed in the "Window" menu and instead the title the window originally had is displayed. Reusing objects is recommended by Apple. I searched the Internet for documentation, examples, or best practices on reusing windows and came up empty.

I tried resetting the window's title to "Window", the title a `NSWindow` has by default, before the window is reused and the "Window" menu displayed the correct title.

The change in the pull request adds the following code to the start of `MainWindowController.windowWillOpen`:
```swift
// The titles of open windows shown in the "Window" menu are automatically managed by the AppKit
// framework. To improve performance PlayerCore caches and reuses player instances along with
// their windows. If a window is reused the framework will display the title first used for the
// window in the "Window" menu even after IINA has updated the title of the window unless the
// window's title is reset to the default title "Window" before it is reusued.
window!.title = "Window"
```

Troubles me that I can't point to documentation explaining why this works.

**Comment by saagarjha** [comment]
I'm unable to reproduce this problem on macOS Monterey 12.3 Beta (21E5206e). Is it still occurring? If so, are the steps to reproduce still the same?

**Comment by russell-kitchen** [comment]
This issue is still occurring for me on iina 1.2.0 build 129, Big Sur 11.5.2

**Comment by low-batt** [comment]
I just now built from the develop branch and tested under macOS 12.2.1 and could not reproduce the problem. I took that same executable to macOS 10.15.7 and reproduced the problem. Thanks @russell-kitchen for confirming it reproduces under macOS 11. This did seem like a bug in AppKit. Looks like Apple has fixed it in Monterey.

I will update the fix to document it as a workaround for an AppKit defect and use #available to conditionalize it to not apply the workaround from macOS 12 onwards.

**Comment by low-batt** [comment]
I've updated the PR to only apply the workaround in older versions of macOS. I've updated the comments to explain the AppKit defect has been fixed in Monterey and reference the related issues.

**Comment by low-batt** [comment]
The fix (a workaround) in PR #3435 has been merged into the develop branch.

Apple has fixed the AppKit defect in macOS Monterey.

**Comment by low-batt** [comment]
Closing. Fixed in IINA 1.3.0.
